File: src/main/java/com/github/kayjamlang/core/Expression.java
package com.github.kayjamlang.core;

import com.github.kayjamlang.core.opcodes.AccessIdentifier;

import java.util.HashMap;
import java.util.Map;

public abstract class Expression implements Cloneable {
    public final AccessIdentifier identifier;
    public HashMap<String, Object> data = new HashMap<>();
    public final int line;

    public Expression(AccessIdentifier identifier, int line) {
        this.identifier = identifier;
        this.line = line;
    }

    @Override
    public String toString() {
        return "Expression{" +
                "identifier=" + identifier +
                ", line=" + line +
                '}';
    }

    @Override
    @SuppressWarnings("unchecked")
    protected Expression clone() throws CloneNotSupportedException {
        Expression expression = (Expression) super.clone();
        expression.data = (HashMap<String, Object>) data.clone();
        return expression;
    }
}

File: src/main/scala/com/github/kayjamlang/core/expression/Expression.scala
package com.github.kayjamlang.core.expression

import com.github.kayjamlang.core.opcodes.AccessType

import scala.collection.mutable

abstract class Expression (val access: AccessType, val line: Int) extends Cloneable {
  var data = new mutable.HashMap[String, Any]

  @SuppressWarnings(Array("unchecked"))
  @throws[CloneNotSupportedException]
  override protected def clone: Expression = {
    val expression = super.clone.asInstanceOf[Expression]
    expression.data = data clone()
    expression
  }
}